src/Transports.AspNetCore/GraphQLHttpMiddleware.cs

```diff
@@ -1,43 +1,221 @@
+
+    /// <summary>
+    /// Initializes a new instance.
+    /// </summary>
+    public GraphQLHttpMiddleware(
+        RequestDelegate next,
+        IGraphQLTextSerializer serializer,
+        IDocumentExecuter<TSchema> documentExecuter,
+        IServiceScopeFactory serviceScopeFactory,
+        GraphQLHttpMiddlewareOptions options,
+        IServiceProvider provider,
+        IHostApplicationLifetime hostApplicationLifetime)
+        : this(next, serializer, documentExecuter, serviceScopeFactory, options,
+              CreateWebSocketHandlers(serializer, documentExecuter, serviceScopeFactory, provider, hostApplicationLifetime, options))
+    {
+    }
+
+    /// <summary>
+    /// Initializes a new instance.
+    /// </summary>
+    protected GraphQLHttpMiddleware(
+        RequestDelegate next,
+        IGraphQLTextSerializer serializer,
+        IDocumentExecuter<TSchema> documentExecuter,
+        IServiceScopeFactory serviceScopeFactory,
+        GraphQLHttpMiddlewareOptions options,
+        IEnumerable<IWebSocketHandler<TSchema>>? webSocketHandlers = null)
+        : base(next, serializer, options, webSocketHandlers)
+    {
+        _documentExecuter = documentExecuter ?? throw new ArgumentNullException(nameof(documentExecuter));
+        _serviceScopeFactory = serviceScopeFactory ?? throw new ArgumentNullException(nameof(serviceScopeFactory));
+        var getRule = new HttpGetValidationRule();
+        _getValidationRules = DocumentValidator.CoreRules.Append(getRule).ToArray();
+        _getCachedDocumentValidationRules = new[] { getRule };
+        var postRule = new HttpPostValidationRule();
+        _postValidationRules = DocumentValidator.CoreRules.Append(postRule).ToArray();
+        _postCachedDocumentValidationRules = new[] { postRule };
+    }
+
+    private static IEnumerable<IWebSocketHandler<TSchema>> CreateWebSocketHandlers(
+        IGraphQLSerializer serializer,
+        IDocumentExecuter<TSchema> documentExecuter,
+        IServiceScopeFactory serviceScopeFactory,
+        IServiceProvider provider,
+        IHostApplicationLifetime hostApplicationLifetime,
+        GraphQLHttpMiddlewareOptions options)
+    {
+        if (hostApplicationLifetime == null)
+            throw new ArgumentNullException(nameof(hostApplicationLifetime));
+        if (provider == null)
+            throw new ArgumentNullException(nameof(provider));
+
+        return new IWebSocketHandler<TSchema>[] {
+            new WebSocketHandler<TSchema>(serializer, documentExecuter, serviceScopeFactory, options,
+                hostApplicationLifetime, provider.GetService<IWebSocketAuthenticationService>()),
+        };
+    }
+
+    ******************************/
+
+    /// <inheritdoc/>
+    protected override async Task<ExecutionResult> ExecuteScopedRequestAsync(HttpContext context, GraphQLRequest? request, IDictionary<string, object?> userContext)
+    {
+        var scope = _serviceScopeFactory.CreateScope();
```

Comment:
```suggestion
        var scope = context.RequestServices.CreateScope()
```

and remove `_serviceScopeFactory` field ?

Reply:
It will require another call to the resolver at runtime that way, because `.CreateScope()` calls `.GetRequiredService<IServiceScopeFactory>().CreateScope()`.  I would leave it as-is ... wouldn't you?

https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ServiceProviderServiceExtensions.cs#L101-L104

Reply:
It just seems like we will get better performance if we already have references to all singleton services possible.  We certainly could eliminate `_serviceScopeFactory`, `_documentExecuter` and `_serializer`, pulling them from DI at runtime... but I don't think that's best for performance.

Reply:
Instant access time 
![изображение](https://user-images.githubusercontent.com/21261007/169610462-c04101df-3ae4-4b91-96fc-f7c7ba147602.png)

`IServiceScopeFactory` dependency looks excessive. I'm fine to keep other dependencies.

Reply:
Can you explain?  I do not follow.

Reply:
If it takes "work" to construct the service scope factory, then by all means it should be done during initialization.

Reply:
The only reason I would change it is if we wanted to have scopes created as sub-scopes of the request scope (MSDI does not support this):

https://docs.autofac.org/en/latest/lifetime/index.html#scopes-and-hierarchy

Reply:
I mean only that specifically this dependency is somewhat different from the rest in the sense that it is not mandatory and it may not be obtained at all. I understand that one can pull all dependencies from `HttpContext.RequestServices`. I do not call for this - _documentExecuter and _serializer should be taken as dependencies. I'm saying more about not to consider any API for working with scopes as dependency. It's internal infrastructure code.

Reply:
We can change it.  I'll add an issue for this.